Per backend configuration

jefferai opened the pull request:
This adds initial support for per-backend configuration. Initially, this enables overriding the global default lease TTL and max lease TTL, but puts all of the plumbing in place for future enhancements.

BRMatt commented on command/mounttune.go:
```diff
@@ -0,0 +1,106 @@
+	helpText := `
+  Usage: vault mount-tune [options] path
+
+  Tune configuration options fro a mounted secret backend.
```
Looks like `fro` should be `for`

jefferai replied:
Yep. Will fix that tomorrow along with the other things I need to finish. Thanks!

armon commented on vault/mount.go:
```diff
@@ -110,10 +132,17 @@ type MountEntry struct {
 
+// MountConfig is used to hold settable options
+type MountConfig struct {
+	DefaultLeaseTTL *time.Duration `json:"default_lease_ttl" structs:"default_lease_ttl" mapstructure:"default_lease_ttl"` // Override for global default
```
Any reason to make these a pointer to a time.Duration instead of directly embedding the time.Duration?

jefferai replied:
Yes -- because when these are passed around as part of a tune call, I have to know whether the values were unset, or set to 0 -- IOW, use the defaults. If the values are unset I don't want to change them, but the time.Duration zero value is 0.

I could have a separate struct for the http/api/command functions but that seemed unnecessary. When these are created within core, they are initialized to zero.

armon replied:
But the 0 value isn't valid anyways. Can we just treat 0 as unset/use default?

armon commented on vault/mount.go:
```diff
@@ -479,6 +575,85 @@ func (c *Core) newLogicalBackend(t string, view logical.Storage, conf map[string
+}
+
+// mountEntryByPath searches across all tables to find the MountEntry
+func (c *Core) mountEntryByPath(path string) (*MountEntry, error) {
```
I would just add a `MatchingSystemView` to Router like `MatchingView` already exists and do the lookup there. Much more efficient since its already using radix trees to index by path.

jefferai replied:
That's fine, but keep in mind that there is still a need to look up mount entries by path (potentially in the router) anywhere that I need to know if the value is actually the global default or backend-specific -- the sysview only indicates the value that the backend should use, regardless of which source it came from.

armon commented on vault/mount.go:
```diff
@@ -479,6 +575,85 @@ func (c *Core) newLogicalBackend(t string, view logical.Storage, conf map[string
+}
+
+// sysViewByPath is a simple helper for MountEntrySysView
+func (c *Core) sysViewByPath(path string) (logical.SystemView, error) {
```
If you use the `MatchingSystemView` on Router, this can go away.

armon commented:
So I would re-work the flow a little bit to simplify. My recommendation would be to add a `SystemView` to the `Mount` of Router. This way you can do an efficient lookup of path -> SystemView. I'd implement dynamicSystemView with a pointer to the core (to get the global defaults) and a pointer to the override table. This way a bunch of the new `Core` methods go away and simplifies the interacts.

armon commented on vault/mount.go:
```diff
@@ -350,19 +384,80 @@ func (c *Core) remount(src, dst string) error {
+	}
+
+	// Find and modify mount
+	for _, ent := range c.mounts.Entries {
```
We can probably pull this lookup into a helper or just get the Router to give it back to us

armon commented on vault/mount.go:
```diff
@@ -350,19 +384,80 @@ func (c *Core) remount(src, dst string) error {
+	for _, ent := range c.mounts.Entries {
+		if ent.Path == path {
+			if config.MaxLeaseTTL != nil {
+				if *config.MaxLeaseTTL == 0 {
```
Doesn't this logic always set the `et.Config.MaxLeaseTTL` to the `config.MaxLeaseTTL` value? Hard to see in the tree of ifs.

armon commented on vault/mount.go:
```diff
@@ -350,19 +384,80 @@ func (c *Core) remount(src, dst string) error {
+							*config.DefaultLeaseTTL, c.maxLeaseTTL)
+					}
+				} else {
+					if *ent.Config.MaxLeaseTTL != 0 && *ent.Config.MaxLeaseTTL < *config.DefaultLeaseTTL {
```
Lets move this validation to be way earlier

jefferai replied:
I'm not sure what you mean here, but the Max logic has to come first, so that if the value is set in the call, that updated value is used when evaluating the Default logic.

armon commented on vault/mount.go:
```diff
@@ -350,19 +384,80 @@ func (c *Core) remount(src, dst string) error {
+			}
+			if config.DefaultLeaseTTL != nil {
+				if *ent.Config.MaxLeaseTTL == 0 {
+					if *config.DefaultLeaseTTL > c.maxLeaseTTL {
```
Is it an issue for default of a backend (e.g. PKI) to be greater than max of global? I'm thinking global max 7 days, PKI default maybe 30days or something

jefferai replied:
It is an issue, which is why it's an error here. If the backend doesn't have a max value overriding the global value, the backend default value cannot be greater than the global max.

armon commented on vault/router.go:
```diff
@@ -91,6 +91,15 @@ func (r *Router) Remount(src, dst string) error {
 
 	// Update the mount point
 	r.root.Delete(src)
+	mountEntry, ok := raw.(*mountEntry)
```
No need to guard the cast here since we control the inputs. We should maybe make a private `remount` method on `dynamicSystemView` and just invoke that to encapsulate the logic.

jefferai replied:
We could, although it doesn't buy us much -- we'd still have to ensure that it is a dynamicSystemView, and if so, invoke remount(newpath) vs dynSysView.path = newpath.

armon commented on vault/core.go:
```diff
@@ -477,14 +477,31 @@ func (c *Core) handleRequest(req *logical.Request) (retResp *logical.Response, r
 	if resp != nil && resp.Secret != nil && !strings.HasPrefix(req.Path, "sys/renew/") {
+		// Get the SystemView for the mount
+		sysView, err := c.sysViewByPath(req.Path)
+		if err != nil {
```
It seems strange that this path could ever result in an error... The log statements should probably be more useful like "[ERR] core: blah blah: err"

jefferai replied:
The log statements are already in that format, so they're just written verbatim here to avoid [ERR] core: [ERR] core: ...

As for whether the path could ever result in an error, the lookup functions that the path is being sent to return errors, so I deferred to that (for the reasons above).

armon commented on logical/system_view.go:
```diff
@@ -6,12 +6,12 @@ import "time"
 type SystemView interface {
 	// DefaultLeaseTTL returns the default lease TTL set in Vault configuration
-	DefaultLeaseTTL() time.Duration
+	DefaultLeaseTTL() (time.Duration, error)
```
We should just ensure the error is not possible, I don't really see how it could happen given we are referencing values we have in-memory.

armon replied:
Otherwise its rather bizarre to handle both for core and for the backend itself

jefferai replied:
Errors happen if the path is invalid. When the lookup function for the path is called, even if we move to using the router, it can return an error. The only way to not expose that error would be to return default duration values (e.g. 0) -- but since that's not actually valid, backends then need to know that they need to check for this condition and do something about it. So it just obfuscates the error condition and potentially leads to incorrect behavior.

armon commented on command/mounttune.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/hashicorp/vault/vault"
+)
+
+// RemountCommand is a Command that remounts a mounted secret backend
```
Copy/paste typo!

armon commented on command/mounttune.go:
```diff
@@ -0,0 +1,106 @@
+	var defaultLeaseTTL, maxLeaseTTL string
+	flags := c.Meta.FlagSet("mount-tune", FlagSetDefault)
+	flags.StringVar(&defaultLeaseTTL, "default-lease-ttl", "", "")
+	flags.StringVar(&maxLeaseTTL, "max-lease-ttl", "", "")
```
Can we have a `-set` flag that switches from reading the tune values (default) to setting the tune values (with -set)? Is there some other way to query the tune values?

jefferai replied:
I didn't implement a read on the CLI mount-tune command for two reasons: one, the info is shown in "mounts", and two, because due to the information that mount-tune is looking up you always get the current values (IOW, it's pulling the information from a SysView, not from the mount entries), not a "0" for the system default. In our previous conversation we decided to return 0 when the system default is being used. Otherwise you can get something back like a default of 180s and a max of 60s. Lease TTLs will be limited to the backend max, but it's showing the correct system default.

armon replied:
Makes sense! Didn't realize that info was in "mounts"

jefferai commented:
> I'd implement dynamicSystemView with a pointer to the core (to get the global defaults) and a pointer to the override table. This way a bunch of the new Core methods go away and simplifies the interacts.

I don't follow what you mean here in regards to an override table. I could add a table that maps a path to a particular MountEntry's Config object. That table would still have to be updated on any change to the mount table. Same with putting a pointer to the MountEntry's Config object directly in the dynamicSystemView -- I could do that and move logic there, but it does still require logic to handle remounts.

armon commented on website/source/docs/http/sys-mounts.html.md:
````diff
@@ -31,12 +36,54 @@ description: |-
+
+    ```javascript
+    {
+      "config": {
````
Any reason to nest under "config" instead of making those keys top-level? Do we anticipate another block besides "config"?

jefferai replied:
I did that purely to keep it consistent with mounts, but I guess that doesn't make sense since what it returns (the systemview values instead of the mount entry values) is different. So I'll change that.

armon commented on vault/mount.go:
```diff
@@ -323,6 +330,7 @@ func (c *Core) remount(src, dst string) error {
 	// Update the entry in the mount table
 	newTable := c.mounts.Clone()
 	for _, ent := range newTable.Entries {
+		//FIXME: Update systemview in each logical backend
```
Router does this it looks like? So nothing to fix

jefferai replied:
Ah, it does now, yes.

armon commented on vault/dynamic_system_view.go:
```diff
@@ -0,0 +1,54 @@
+	def = d.core.defaultLeaseTTL
+	max = d.core.maxLeaseTTL
+
+	if me.Config.DefaultLeaseTTL != nil && *me.Config.DefaultLeaseTTL != 0 {
```
So we treat nil and 0 as default case anyways, so we may as well remove it being a pointer.

armon commented on vault/dynamic_system_view.go:
```diff
@@ -0,0 +1,54 @@
+
+type dynamicSystemView struct {
+	core *Core
+	path string
```
Can we just put the *MountEntry in here directly and thus avoid the call to `d.core.router.MatchingMountEntry`?

jefferai commented:
@armon I think I've addressed all of your feedback. Take a look at what I did with Clone() -> ShallowClone() -- it's very straightforward and it ensures that the data seen by the pointers stays consistent
